gameofthrones/src/MainPage.js

```diff
@@ -0,0 +1,140 @@
+                isLoading : false,
+                hasError : true
+            });
+        });
+    }
+    
+
+    componentDidMount() {
+        this.getFirstPage();        
+    }
+
+    getNextPage = ()=> 
+    {
+        let currentpage = this.state.pageNumber + 1;
+        
+        fetch('https://www.anapioficeandfire.com/api/characters?page='+currentpage+'&pageSize=50')
+        .then( response => {                
+                return (response.json());
+        })
+        .then(data => {
+            this.setState({
+                characters : data,
+                pageNumber : currentpage,
+                isLoading : false
+            });
+        })
+        .catch(error=>{
+            console.log(error.message);
+            this.setState({
+                isLoading : false,
+                hasError : true
+            });
+        });
+    }
+
+    getPreviousPage = ()=> 
+    {
+        let currentpage = this.state.pageNumber;
+        if (currentpage > 1)
+        {
+            currentpage = currentpage - 1;
+        }
+        
+        fetch('https://www.anapioficeandfire.com/api/characters?page='+currentpage+'&pageSize=50')
+        .then( response => {                
+                return (response.json());
+        })
+        .then(data => {            
+            this.setState({
+                characters : data,
+                pageNumber : currentpage,
+                isLoading : false
+            });
+        })
+        .catch(error=>{
+            console.log(error.message);
+            this.setState({
+                isLoading : false,
+                hasError : true
+            });
+        });
+    }
+
+    backToHome= () => {
+        this.props.history.push('/');
+    }
+
+    displayCharacterDetails = () => {        
+        return(            
+            <Character selectedItem={this.state.selectedCharacter} onClick = {this.backToHome} />
+        )
+    }
+
+    navigateToCharacterDetails = (character) => {
+        this.setState(prevState => {
+            return {
+                selectedCharacter : character                
+            };
+        })
+        this.props.history.push('/character/');
+    }
+
+    render() {       
+        if (this.state.isLoading) {
+            return <div>Loading...</div>;
+        }
+
+        if (this.state.hasError) {
+            return <div>ERROR, please reload and try again</div>;
+        }
+
+        return (            
+            <div>
+                <Switch>
+                    <Route exact path="/" render={(props) => 
+                        <Directory  characters={this.state.characters}
+                                     onClick = {this.navigateToCharacterDetails} 
+                                     getNextPage = {this.getNextPage}
+                                     getPreviousPage = {this.getPreviousPage}/>} />
+                    <Route path="/character" render={this.displayCharacterDetails} />
```

Comment:
By passing character data inside the props of the component instead of as a url parameter you are missing out on much of the power of using react router. 
For example all your character detail pages have the url http://localhost:3000/character/
but if you were to paste http://localhost:3000/character/ directly into the url your app will crash
instead pass some sort of character id that initiates a fetch request in the character component.

Reply:
instead have <Route path="/character/:id" render={this.displayCharacterDetails} />